services/repository/files/tree.go

```diff
@@ -118,3 +122,389 @@ func GetTreeBySHA(ctx context.Context, repo *repo_model.Repository, gitRepo *git
 	}
 	return tree, nil
 }
+
+func entryModeString(entryMode git.EntryMode) string {
+	switch entryMode {
+	case git.EntryModeBlob:
+		return "blob"
+	case git.EntryModeExec:
+		return "exec"
+	case git.EntryModeSymlink:
+		return "symlink"
+	case git.EntryModeCommit:
+		return "commit" // submodule
+	case git.EntryModeTree:
+		return "tree"
+	}
+	return "unknown"
+}
+
+type TreeViewNode struct {
+	Name         string          `json:"name"`
+	Type         string          `json:"type"`
+	Path         string          `json:"path"`
+	SubModuleURL string          `json:"sub_module_url,omitempty"`
+	Children     []*TreeViewNode `json:"children,omitempty"`
+}
+
+func (node *TreeViewNode) sortLevel() int {
+	switch node.Type {
+	case "tree", "commit":
+		return 0
+	default:
+		return 1
+	}
+}
+
+func newTreeViewNodeFromEntry(ctx context.Context, commit *git.Commit, parentDir string, entry *git.TreeEntry) *TreeViewNode {
+	node := &TreeViewNode{
+		Name: entry.Name(),
+		Type: entryModeString(entry.Mode()),
+		Path: path.Join(parentDir, entry.Name()),
+	}
+
+	if node.Type == "commit" {
+		if subModule, err := commit.GetSubModule(node.Path); err != nil {
+			log.Error("GetSubModule: %v", err)
+		} else if subModule != nil {
+			submoduleFile := git.NewCommitSubmoduleFile(subModule.URL, entry.ID.String())
+			webLink := submoduleFile.SubmoduleWebLink(ctx)
+			node.SubModuleURL = webLink.CommitWebLink
+		}
+	}
+
+	return node
+}
+
+// sortTreeViewNodes list directory first and with alpha sequence
+func sortTreeViewNodes(nodes []*TreeViewNode) {
+	sort.Slice(nodes, func(i, j int) bool {
+		if nodes[i].sortLevel() != nodes[j].sortLevel() {
+			return nodes[i].sortLevel() < nodes[j].sortLevel()
+		}
+		return nodes[i].Name < nodes[j].Name
+	})
+}
+
+/*
+Example 1: (path: /)
+
+	GET /repo/name/tree/
+
+	resp:
+	[{
+	    "name": "d1",
+	    "isFile": false,
+	    "path": "d1"
+	},{
+	    "name": "d2",
+	    "isFile": false,
+	    "path": "d2"
+	},{
+	    "name": "d3",
+	    "isFile": false,
+	    "path": "d3"
+	},{
+	    "name": "f1",
+	    "isFile": true,
+	    "path": "f1"
+	},]
+
+Example 2: (path: d3)
+
+	GET /repo/name/tree/d3
+	resp:
+	[{
+	    "name": "d3d1",
+	    "isFile": false,
+	    "path": "d3/d3d1"
+	}]
+
+Example 3: (path: d3/d3d1)
+
+	GET /repo/name/tree/d3/d3d1
+	resp:
+	[{
+	    "name": "d3d1f1",
+	    "isFile": true,
+	    "path": "d3/d3d1/d3d1f1"
+	},{
+	    "name": "d3d1f1",
+	    "isFile": true,
+	    "path": "d3/d3d1/d3d1f2"
+	}]
+*/
+func GetTreeList(ctx context.Context, repo *repo_model.Repository, gitRepo *git.Repository, treePath string, ref git.RefName, recursive bool) ([]*TreeViewNode, error) {
+	if repo.IsEmpty {
+		return nil, nil
+	}
+	if ref == "" {
+		ref = git.RefNameFromBranch(repo.DefaultBranch)
+	}
+
+	// Check that the path given in opts.treePath is valid (not a git path)
+	cleanTreePath := CleanUploadFileName(treePath)
```

Comment:
It is not "uploading", why `CleanUploadFileName`?

Reply:
4eac502918